crate/server/src/routes/findex.rs

```diff
@@ -0,0 +1,142 @@
+
+    let bytes_slice = bytes.as_ref();
+    let addresses = Addresses::deserialize(bytes_slice.to_vec())?;
+
+    trace!(
+        "batch_read: number of addresses {}:",
+        addresses.clone().into_inner().len()
+    );
+
+    let result_words =
+        OptionalWords::new(findex_server.db.batch_read(addresses.into_inner()).await?);
+    //  findex_server.db.batch_read(addresses.into_inner()).awai
+    trace!(
+        "batch_read successful. Number of non null words: {}.",
+        result_words
+            .clone()
+            .into_inner()
+            .iter()
+            .fold(0, |acc, x| acc + (x.is_some() as usize))
+    );
+
+    // Convert Vec<Option<[u8; WORD_LENGTH]>> to Vec<u8>
+    let response_bytes = Bytes::from(
+        result_words.serialize()?, // result_words
+                                   //     .into_iter()
+                                   //     .flat_map(|opt_word| {
+                                   //         let mut bytes = vec![0u8; 1]; // Option discriminant byte
+                                   //         match opt_word {
+                                   //             Some(word) => {
+                                   //                 bytes[0] = 1;
+                                   //                 bytes.extend_from_slice(&<[u8; WORD_LENGTH]>::from(word));
+                                   //             }
+                                   //             None => {
+                                   //                 bytes[0] = 0;
+                                   //             }
+                                   //         }
+                                   //         bytes
+                                   //     })
+                                   //     .collect::<Vec<u8>>(),
+    );
+
+    Ok(HttpResponse::Ok()
+        .content_type("application/octet-stream")
+        .body(Bytes::from(response_bytes)))
+}
+
+#[post("/indexes/{index_id}/guarded_write")]
+pub(crate) async fn findex_guarded_write(
+    req: HttpRequest,
+    index_id: web::Path<String>,
+    bytes: Bytes,
+    findex_server: Data<Arc<FindexServer>>,
+) -> ResponseBytes {
+    let user = findex_server.get_user(&req);
+    info!("user {user}: POST /indexes/{index_id}/guarded_write");
+
+    check_permission(&user, &index_id, Permission::Write, &findex_server).await?;
+
+    let discriminant_flag = bytes[ADDRESS_LENGTH]; // 0 or 1. 0 means None, 1 means Some. Assumes the first ADDRESS_LENGTH bytes are the address
+    assert!(
+        discriminant_flag == 0 || discriminant_flag == 1,
+        "Invalid discriminant flag. Expected 0 or 1, found {}",
+        discriminant_flag
+    );
+    let guard_len = if discriminant_flag == 0 {
+        ADDRESS_LENGTH + 1
+    } else {
+        ADDRESS_LENGTH + 1 + WORD_LENGTH
+    };
+    let guard = Guard::deserialize(bytes[..guard_len].to_vec())?;
+    let tasks = Tasks::deserialize(bytes[guard_len..].to_vec())?;
```

Comment:
This seems too verbose but it's necessary otherwise the server is almost impossible to debug in case of error. It also won't pass the CI if these checks are not made. This is just a comment, mark solved if it opened an issue.